File: src/git_forge_release_migrator/core/versioning.py
def _normalize(tag: str) -> tuple[int, int, int]:
    value = tag.strip()
    if value.startswith("v"):
        value = value[1:]
    parts = value.split(".")
    if len(parts) != 3:
        raise ValueError(f"Invalid semantic tag: {tag}")
    return int(parts[0]), int(parts[1]), int(parts[2])


def version_le(left: str, right: str) -> bool:
    return _normalize(left) <= _normalize(right)

File: dart_cli/lib/src/core/versioning.dart
final class SemverUtils {
  const SemverUtils._();

  static List<int> _normalize(String tag) {
    String value = tag.trim();
    if (value.startsWith('v')) {
      value = value.substring(1);
    }

    final List<String> parts = value.split('.');
    if (parts.length != 3) {
      throw ArgumentError('Invalid semantic tag: $tag');
    }

    try {
      return <int>[
        int.parse(parts[0]),
        int.parse(parts[1]),
        int.parse(parts[2]),
      ];
    } on FormatException {
      throw ArgumentError('Invalid semantic tag: $tag');
    }
  }

  static bool versionLe(String left, String right) {
    final List<int> normalizedLeft = _normalize(left);
    final List<int> normalizedRight = _normalize(right);
    for (int index = 0; index < 3; index += 1) {
      if (normalizedLeft[index] < normalizedRight[index]) {
        return true;
      }

      if (normalizedLeft[index] > normalizedRight[index]) {
        return false;
      }
    }

    return true;
  }
}
